cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +562,514 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
+    }
+    return query::clustering_range(intersection_start, intersection_end);
+}
+
+/// An expression visitor that translates multi-column atoms into clustering ranges.
+struct multi_column_range_accumulator {
+    const query_options& options;
+    const schema_ptr schema;
+    std::vector<query::clustering_range> ranges{query::clustering_range::make_open_ended_both_sides()};
+    const clustering_key_prefix::prefix_equal_tri_compare prefix3cmp = get_unreversed_tri_compare(*schema);
+
+    void operator()(const binary_operator& binop) {
+        if (auto tup = dynamic_pointer_cast<tuples::value>(binop.rhs->bind(options))) {
+            if (!is_compare(binop.op)) {
+                on_internal_error(
+                        rlogger, format("multi_column_range_accumulator: unexpected atom {}", binop));
+            }
+            auto opt_values = tup->get_elements();
+            auto& lhs = std::get<std::vector<column_value>>(binop.lhs);
+            std::vector<bytes> values(lhs.size());
+            for (size_t i = 0; i < lhs.size(); ++i) {
+                values[i] = deref_column_value(opt_values[i], lhs.at(i).col->name_as_text());
+            }
+            intersect_all(to_range(binop.op, clustering_key_prefix(values)));
+        } else if (auto dv = dynamic_pointer_cast<lists::delayed_value>(binop.rhs)) {
+            must_be_in_operator(binop);
+            process_in_values(
+                    dv->get_elements() | transformed(
+                            [&] (const ::shared_ptr<term>& t) {
+                                return static_pointer_cast<tuples::value>(t->bind(options))->get_elements();
+                            }));
+        } else if (auto mkr = dynamic_pointer_cast<tuples::in_marker>(binop.rhs)) {
+            // This is `(a,b) IN ?`.  RHS elements are themselves tuples, represented as vector<bytes_opt>.
+            must_be_in_operator(binop);
+            process_in_values(
+                    static_pointer_cast<tuples::in_value>(mkr->bind(options))->get_split_values());
+        }
+    }
+
+    void operator()(const conjunction& c) {
+        std::ranges::for_each(c.children, [this] (const expression& child) { std::visit(*this, child); });
+    }
+
+    void operator()(bool b) {
+        if (!b) {
+            ranges.clear();
+        }
+    }
+
+    static void must_be_in_operator(const binary_operator& binop) {
+        if (binop.op != oper_t::IN) {
+            on_internal_error(rlogger, format("multi_column_range_accumulator: unexpected atom {}", binop));
+        }
+    }
+
+    /// Intersects each range with v.  If any intersection is empty, clears ranges.
+    void intersect_all(const query::clustering_range& v) {
+        for (auto& r : ranges) {
+            auto intrs = intersection(r, v, prefix3cmp);
+            if (!intrs) {
+                ranges.clear();
+                break;
+            }
+            r = *intrs;
+        }
+    }
+
+    template<std::ranges::range Range>
```

Comment:
Does clang not barf on range constraints? It did for me last time I checked.

Reply:
And why is this a template in the first place, isn't it already specialized for clustering keys?

Reply:
> Does clang not barf on range constraints?

I only saw it barf on std::range of vector<vector<...>>.  Everything else I tried compiles just fine.

> It did for me last time I checked.

I doubt that, having seen [this](https://github.com/scylladb/scylla/pull/8208/files#diff-f112029dbdd3dc880fe432ade591c2f50fc777491902d6147cd45e37baa62380R75).

> And why is this a template in the first place, isn't it already specialized for clustering keys?

It gets invoked on `transformed` results.